wagtail/query.py

```diff
@@ -243,17 +243,29 @@ def public_q(self):
             q &= ~self.descendant_of_q(restriction.page, inclusive=True)
         return q
 
+    def private_q(self):
+        from wagtail.models import PageViewRestriction
+
+        q = Q()
+        for restriction in PageViewRestriction.objects.select_related("page").all():
+            q |= self.descendant_of_q(restriction.page, inclusive=True)
+
+        # do not match any page if no private section exists.
+        return q if q else Q(pk__in=[])
+
     def public(self):
         """
-        This filters the QuerySet to only contain pages that are not in a private section
+        This filters the QuerySet to only contain pages that are not in a private
+        section and their descendants.
         """
-        return self.filter(self.public_q())
+        return self.exclude(self.private_q())
 
     def not_public(self):
```

Comment:
Question: Should we also have a `private` filter method here too? It'd do exactly the same thing, but be much clearer to someone reading the code what's happening. Would have to be in addition to `not_public` to avoid a breaking change.

Reply:
Yes, I have also been thinking about that. The name not_public is not really clear. Should I add it to this PR, or open a new one?

Reply:
Doing it here is probably fine, just putting `private = not_public` underneath this method _should_ do it (plus accompanying documentation and tests)

Reply:
I believe it is now ready to be merged. I also added the regression test for the bug.